### PR TITLE
ENH: Improve api.types

### PR DESCRIPTION
**Pull request by bashtage**
Clean and improve the typing for api.types
Add tests

- [X] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

**Comment by bashtage** [comment]
xref pandas-dev/pandas#48410

**Comment by bashtage** [comment]
Removed 2 deprecated `is_` functions.
Removed some `import x as x` import that were only reexports

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: pandas-stubs/api/types/__init__.pyi [hunk 1]
```diff
@@ -4,7 +4,6 @@ from pandas.core.dtypes.api import (
     is_array_like as is_array_like,
     is_bool as is_bool,
     is_bool_dtype as is_bool_dtype,
-    is_categorical as is_categorical,
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/api/types/__init__.pyi:7]: is_categorical is deprecated in favor of is_categorical_dtype.

File: pandas-stubs/api/types/__init__.pyi [hunk 2]
```diff
@@ -15,7 +14,6 @@ from pandas.core.dtypes.api import (
     is_dict_like as is_dict_like,
     is_dtype_equal as is_dtype_equal,
     is_extension_array_dtype as is_extension_array_dtype,
-    is_extension_type as is_extension_type,
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/api/types/__init__.pyi:18]: is_extension_type is deprecated in favor of is_extension_array_type

File: pandas-stubs/api/types/__init__.pyi (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas-stubs/api/types/__init__.pyi:7]: Looks like the deprecation is not documented.  Can you create an issue in `pandas` about the missing doc?

File: pandas-stubs/api/types/__init__.pyi (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply pandas-stubs/api/types/__init__.pyi:18]: OK - that one is documented!

File: pandas-stubs/api/types/__init__.pyi (hunk 1 above)
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/api/types/__init__.pyi:7]: PR opened

File: pandas-stubs/_libs/lib.pyi [hunk 3]
```diff
@@ -1,3 +1,15 @@
...
 def infer_dtype(value: object, skipna: bool = ...) -> str: ...
+def is_iterator(obj: object) -> bool: ...
+def is_scalar(val: object) -> bool: ...
+
+# TODO: Add to pandas docs
+def is_decimal(val: object) -> bool: ...
+def is_list_like(obj: object, allow_sets: bool = ...) -> bool: ...
+
+# TODO: Add to pandas docs
+def is_period(val: object) -> bool: ...
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/_libs/lib.pyi]: Removed these. Shouldn't be there or exposed in pandas.

File: tests/test_api_types.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+import pandas as pd
+from typing_extensions import assert_type
+
+from pandas._typing import DtypeObj
+
+import pandas.core.dtypes.api as api
```
[the remark this answers is not in the document]
Comment (bashtage) [reply tests/test_api_types.py]: All tested now.

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**